### PR TITLE
[SPARK-38503][K8S] Warn if `VolcanoFeatureStep.getAdditionalPreKubernetesResources` is used in executor side

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
- Add warn for getAdditionalPreKubernetesResources in executor side

### Why are the changes needed?
Only `spark.kubernetes.scheduler.volcano.podGroupTemplateFile` is going to be supported in v3.3, and it only be called when building driver side feature steps. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- K8S IT passed
- CI passed

### CODE REVIEW

**Comment by Yikun** [comment]
cc @dongjoon-hyun

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

File: docs/running-on-kubernetes.md [hunk 1]
```diff
@@ -1547,6 +1547,22 @@ See the [configuration page](configuration.html) for information on Spark config
...
+<tr>
+  <td><code>spark.kubernetes.driver.podGroupTemplateFile</code></td>
+  <td>(none)</td>
+  <td>
+    File containing a template pod group spec for driver.
+  </td>
+  <td>3.3.0</td>
+</tr>
+<tr>
+  <td><code>spark.kubernetes.executor.podGroupTemplateFile</code></td>
```
[the remark this answers is not in the document]
Comment (Yikun) [reply docs/running-on-kubernetes.md:1559]: OK, will address. I think executor side separated podgroup support will not be delivered in v3.3 as original plan.

**Comment by Yikun** [comment]
```
[info] KubernetesSuite:
[info] - Run SparkPi with no resources (9 seconds, 179 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (8 seconds, 592 milliseconds)
[info] - Run SparkPi with a very long application name. (8 seconds, 593 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (8 seconds, 579 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (8 seconds, 682 milliseconds)
[info] - Run SparkPi with an argument. (8 seconds, 600 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (8 seconds, 604 milliseconds)
[info] - All pods have the same service account by default (9 seconds, 567 milliseconds)
[info] - Run extraJVMOptions check on driver (4 seconds, 481 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (8 seconds, 623 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (16 seconds, 299 milliseconds)
[info] - Run SparkPi with env and mount secrets. (17 seconds, 24 milliseconds)
[info] - Run PySpark on simple pi.py example (9 seconds, 572 milliseconds)
[info] - Run PySpark to test a pyfiles example (11 seconds, 669 milliseconds)
[info] - Run PySpark with memory customization (9 seconds, 583 milliseconds)
[info] - Run in client mode. (7 seconds, 336 milliseconds)
[info] - Start pod creation from template (9 seconds, 692 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (9 seconds, 666 milliseconds)
[info] - Test basic decommissioning (42 seconds, 9 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (42 seconds, 465 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 41 seconds)
[info] - Test decommissioning timeouts (41 seconds, 795 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (8 minutes, 27 seconds)
[info] VolcanoSuite:
[info] - Run SparkPi with no resources (11 seconds, 119 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (10 seconds, 594 milliseconds)
[info] - Run SparkPi with a very long application name. (10 seconds, 582 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (10 seconds, 586 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (10 seconds, 563 milliseconds)
[info] - Run SparkPi with an argument. (10 seconds, 626 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (9 seconds, 601 milliseconds)
[info] - All pods have the same service account by default (10 seconds, 593 milliseconds)
[info] - Run extraJVMOptions check on driver (5 seconds, 617 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (10 seconds, 800 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (16 seconds, 217 milliseconds)
[info] - Run SparkPi with env and mount secrets. (18 seconds, 401 milliseconds)
[info] - Run PySpark on simple pi.py example (10 seconds, 574 milliseconds)
[info] - Run PySpark to test a pyfiles example (12 seconds, 717 milliseconds)
[info] - Run PySpark with memory customization (10 seconds, 617 milliseconds)
[info] - Run in client mode. (7 seconds, 239 milliseconds)
[info] - Start pod creation from template (11 seconds, 742 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (9 seconds, 691 milliseconds)
[info] - Test basic decommissioning (44 seconds, 866 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (44 seconds, 124 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 42 seconds)
[info] - Test decommissioning timeouts (43 seconds, 717 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 7 seconds)
[info] - Run SparkPi with volcano scheduler (10 seconds, 660 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (31 seconds, 601 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (30 seconds, 562 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (13 seconds, 364 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (25 seconds, 364 milliseconds)
[info] - SPARK-38423: Run SparkPi Jobs with priorityClassName (19 seconds, 300 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (16 seconds, 464 milliseconds)
[info] Run completed in 29 minutes, 6 seconds.
[info] Total number of tests run: 53
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 53, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 1829 s (30:29), completed 2022-3-10 11:09:53
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 2]
```diff
@@ -44,12 +44,8 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
...
   override def getAdditionalPreKubernetesResources(): Seq[HasMetadata] = {
     val client = new DefaultVolcanoClient
-
-    val template = if (kubernetesConf.isInstanceOf[KubernetesDriverConf]) {
-      kubernetesConf.get(KUBERNETES_DRIVER_PODGROUP_TEMPLATE_FILE)
-    } else {
-      kubernetesConf.get(KUBERNETES_EXECUTOR_PODGROUP_TEMPLATE_FILE)
-    }
+    // Currently only driver/job-level PodGroup is supported
+    val template = kubernetesConf.get(KUBERNETES_DRIVER_PODGROUP_TEMPLATE_FILE)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:48]: This will work incorrectly because this class uses `with KubernetesExecutorCustomFeatureConfigStep`. I'd suggest to throw exception in the previous `if .. else ..` statement structure.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 3]
```diff
@@ -59,6 +55,7 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
 
     var spec = pg.getSpec
     if (spec == null) spec = new PodGroupSpec
+    // Overwrite by driver `pod.spec.priorityClassName`
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:58]: Hmm. Do you use two set of priorityClass? In other words, does your system have the following separately?
- priority classes for Pod
- priority classes for PodGroup (volcano)

File: docs/running-on-kubernetes.md [hunk 4]
```diff
@@ -1538,6 +1538,14 @@ See the [configuration page](configuration.html) for information on Spark config
...
   <td>3.3.0</td>
 </tr>
+<tr>
+  <td><code>spark.kubernetes.driver.podGroupTemplateFile</code></td>
+  <td>(none)</td>
+  <td>
+    File containing a template pod group spec for driver.
+  </td>
+  <td>3.3.0</td>
+</tr>
```
Comment (dongjoon-hyun) [inline docs/running-on-kubernetes.md:1548]: It would be better if you put this after `scheduler` configurations.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 5]
```diff
@@ -299,13 +299,6 @@ private[spark] object Config extends Logging {
       .stringConf
       .createOptional
 
-  val KUBERNETES_EXECUTOR_PODGROUP_TEMPLATE_FILE =
-    ConfigBuilder("spark.kubernetes.executor.podGroupTemplateFile")
-      .doc("File containing a template pod group spec for executors")
-      .version("3.3.0")
-      .stringConf
-      .createOptional
-
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:308]: Well, if you want to delete this in this PR, we should wait until next week.

**Comment by dongjoon-hyun** [comment]
We had better have another JIRA for this PR because this removes a conf, @Yikun .

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala (hunk 3 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:58]: For type of priority class, it's the same thing, pod/pg are share the [priorityClass](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass) in Kubernetes.

Most of users are using same priorities series with pod to make sure job and pod scheduling can be considered in a whole.

But I also heard some users want to set a separate priority classes for PodGroup. This is also why I raised my concern before on https://github.com/apache/spark/pull/35776#discussion_r822247430 .

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala (hunk 2 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:48]: This method not be called in executor as we know, but to protected something in future and more strictly, right?

**Comment by Yikun** [comment]
@dongjoon-hyun Yes for sure.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:58]: Ya, that's the reason why I asked you again about that.
1. Do you know the background of separate priority classes for PodGroup?
2. If all custom schedulers really want to use their own priorityClasses, we need to adjust our implementation a little again.

Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:58]: > Do you know the background of separate priority classes for PodGroup?

- A case I heard before is [multiple sheduler](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-multiple-schedulers/), for some reason they want to separate priority class series for differet scheduler. So they want to a separate priority for each scheudlers.
- The other case is pod level priority in same podgroup, they can set different priority for pod. But I think it is not for spark single driver case.

These cases are not very common, but I think the biggest concern in here is simpfied configuration, such as if users want to use volcano priority scheduling in a queue, they have to specify two configuration: the pod template to set priority and the podgroup template to set queue. Generally speaking, they prefer to accept only maintain a podgroup template for resource level scheduling configuration.

So overwrite default priority if user specify priority in podgroup template might be a alternate choice.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:64]: Ya, that sounds fair. I also support more isolations, @Yikun .

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:64]: Here is the PR, https://github.com/apache/spark/pull/35807 .

**Comment by dongjoon-hyun** [comment]
In total, I made two PRs.
- https://github.com/apache/spark/pull/35807
- https://github.com/apache/spark/pull/35809

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 6]
```diff
@@ -44,11 +45,12 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
 
   override def getAdditionalPreKubernetesResources(): Seq[HasMetadata] = {
     val client = new DefaultVolcanoClient
-
+    // Currently only driver/job-level PodGroup is supported
     val template = if (kubernetesConf.isInstanceOf[KubernetesDriverConf]) {
       kubernetesConf.get(KUBERNETES_DRIVER_PODGROUP_TEMPLATE_FILE)
     } else {
-      kubernetesConf.get(KUBERNETES_EXECUTOR_PODGROUP_TEMPLATE_FILE)
+      throw new SparkException("Get additional pre Kubernetes resources is not supported for " +
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:52]:
```suggestion
      throw new SparkException("VolcanoFeatureStep#getAdditionalPreKubernetesResources() is not supported for " +
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 7]
```diff
@@ -41,7 +42,15 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
 
   override def getAdditionalPreKubernetesResources(): Seq[HasMetadata] = {
     val client = new DefaultVolcanoClient
-    val template = kubernetesConf.getOption(POD_GROUP_TEMPLATE_FILE_KEY)
+
+    // Currently only driver/job-level PodGroup is supported
+    val template = if (kubernetesConf.isInstanceOf[KubernetesDriverConf]) {
+      kubernetesConf.getOption(POD_GROUP_TEMPLATE_FILE_KEY)
+    } else {
+      throw new SparkException("VolcanoFeatureStep#getAdditionalPreKubernetesResources() is " +
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:50]: In this case, `Warning` and `No-Op` will be better.

Comment (martin-g) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:50]: Agreed!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 8]
```diff
@@ -41,7 +42,15 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
...
+        "not supported for executor.")
+    }
+
     val pg = template.map(client.podGroups.load(_).get).getOrElse(new PodGroup())
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:54]: If you want to throw exception, why we don't protect this, @Yikun ?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
First, `getAdditionalPreKubernetesResources` is only used in `KubernetesDriverBuilder`. So, technically, this PR is adding a dead-code which raises Exception. It doesn't give any value in the current status.

Second, if you want to prevent the executors from invoking `getAdditionalPreKubernetesResources` in the future. You should prevent the rest of code in the method, too.

https://github.com/apache/spark/blob/a30575e91e82e9f4394add40f1ba9265eb1f7819/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala#L46-L56

In short, if you want to do this correctly, you should add the exception or warning at the first line of this method. However, in any way, it will be a dead code for now.

File: resource-managers/kubernetes/core/src/test/resources/executor-podgroup-template.yml [hunk 9]
```diff
@@ -1,25 +0,0 @@
...
-#
-apiVersion: scheduling.volcano.sh/v1beta1
-kind: PodGroup
-spec:
-  minMember: 1000
-  minResources:
-    cpu: "4"
-    memory: "16Gi"
-  priorityClassName: executor-priority
-  queue: executor-queue
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/resources/executor-podgroup-template.yml:25]: This is a valid removal.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStepSuite.scala [hunk 10]
```diff
@@ -66,4 +66,14 @@ class VolcanoFeatureStepSuite extends SparkFunSuite {
...
+
+  test("SPARK-38503: raise exception for executor pre resource") {
+    val kubernetesConf = KubernetesTestConf.createExecutorConf(new SparkConf())
+    val step = new VolcanoFeatureStep()
+    step.init(kubernetesConf)
+    val e = intercept[SparkException] {
+      step.getAdditionalPreKubernetesResources()
+    }
+    assert(e.getMessage.contains("is not supported for executor"))
+  }
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStepSuite.scala:78]: I believe this is a little overkill if we are going to do logging simply.

**Comment by Yikun** [comment]
@dongjoon-hyun Thanks for your patient, will add a warn in first line, I think it's fair enough.

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with volcano scheduler (11 seconds, 662 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (33 seconds, 98 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (33 seconds, 717 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (17 seconds, 613 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (32 seconds, 118 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (17 seconds, 899 milliseconds)
[info] Run completed in 2 minutes, 29 seconds.
[info] Total number of tests run: 6
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 6, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 282 s (04:42), completed 2022-3-15 12:30:24
```

**Comment by Yikun** [comment]
To avoid confused hidden change I move execut ut resource cleanup stuff in to a separate PR: https://github.com/apache/spark/pull/35857

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 11]
```diff
@@ -40,6 +41,10 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
   }
 
   override def getAdditionalPreKubernetesResources(): Seq[HasMetadata] = {
+    if (kubernetesConf.isInstanceOf[KubernetesExecutorConf]) {
+      logWarning("VolcanoFeatureStep#getAdditionalPreKubernetesResources() is not supported " +
+        "for executor.")
+    }
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:47]: According to the following PR description, there is no plan for `Volcano` to use `getAdditionalPreKubernetesResources` in the future?
```
Only spark.kubernetes.scheduler.volcano.podGroupTemplateFile is going to be supported in v3.3
```

Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:47]: It only be called when building driver side feature steps.

I plused it. :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala [hunk 12]
```diff
@@ -40,6 +41,10 @@ private[spark] class VolcanoFeatureStep extends KubernetesDriverCustomFeatureCon
   }
 
   override def getAdditionalPreKubernetesResources(): Seq[HasMetadata] = {
+    if (kubernetesConf.isInstanceOf[KubernetesExecutorConf]) {
+      logWarning("VolcanoFeatureStep#getAdditionalPreKubernetesResources() is not supported " +
+        "for executor.")
```
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:46]: this should `return Seq.empty;`, no ?

**Comment by Yikun** [comment]
@dongjoon-hyun It's fine from me. : )

**Review by martin-g — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.4.0.
